src/core/detail/btree_internal.h

```diff
@@ -0,0 +1,458 @@
+//    calculated at run-time and has dedicated accessors (similarly to absl). This allows
+//    dense and efficient representation of tree nodes.
+// 3. We assume that we store small items (8, 16 bytes) which will have a large branching
+//    factor (248/16), meaning the tree will stay shallow even for sizes reaching billion nodes.
+// 4. We do not store parent pointer like in absl tree. Instead we use BTreePath to store
+//    hierarchy of parent nodes. That should reduce our overhead even further by few bits per item.
+// 5. We assume we store POD types - this greatly reduces the complexity of the generics
+//    in the code.
+// 6. We support pmr memory resource. This allows us to use pluggable heaps.
+//
+// TODO: (all the ideas taken from absl implementation)
+//       1. to introduce slices when removing items from the tree (avoid shifts).
+//       2. to avoid merging/rebalancing when removing max/min items from the tree.
+//       3. Small tree optimization: when the tree is small with a single root node, we can
+//          allocate less then 256 bytes (special case) to avoid relative blowups in memory for
+//          small trees.
+
+template <typename Policy> class BTreeNode {
+  using _Key = typename Policy::Key_t;
+
+  friend class ::dfly::BTree<_Key, Policy>;
+
+  BTreeNode(bool leaf) : val_(0) {
+    leaf_ = leaf;
+  }
+
+  static constexpr uint16_t kKeySize = sizeof(_Key);
+  static constexpr uint16_t kTargetNodeSize = 256;
+  static constexpr uint16_t kKeyOffset = sizeof(uint64_t);  // 8 bytes for metadata
+
+ public:
+  static constexpr uint16_t kMaxLeafKeys = (kTargetNodeSize - kKeyOffset) / kKeySize;
+  static constexpr uint16_t kMinLeafKeys = kMaxLeafKeys / 2;
+
+  // internal node:
+  // x slots, (x+1) children: x * kKeySize + (x+1) * sizeof(BTreeNode*) = x * (kKeySize + 8) + 8
+  // x = (kTargetNodeSize - 8 - kKeyOffset) / (kKeySize + 8)
+
+  static constexpr uint16_t kMaxInnerKeys =
+      (kTargetNodeSize - sizeof(void*) - kKeyOffset) / (kKeySize + sizeof(void*));
+  static constexpr uint16_t kMinInnerKeys = kMaxInnerKeys / 2;
+
+  using Key_t = _Key;
+
+  void InitSingle(Key_t key) {
+    SetKey(0, key);
+    num_items_ = 1;
+  }
+
+  Key_t Key(unsigned index) const {
+    Key_t res;
+    memcpy(&res, KeyPtr(index), kKeySize);
+    return res;
+  }
+
+  void SetKey(size_t index, Key_t item) {
+    uint8_t* slot = KeyPtr(index);
+    memcpy(slot, &item, kKeySize);
+  }
+
+  BTreeNode** Children() {
+    // aligned to 8 bytes.
+    constexpr uint16_t kChildOffset = kMaxInnerKeys * kKeySize + kKeyOffset;
+
+    uint8_t* ptr = reinterpret_cast<uint8_t*>(this) + kChildOffset;
+    return reinterpret_cast<BTreeNode**>(ptr);
+  }
+
+  BTreeNode* Child(unsigned i) {
+    return Children()[i];
+  }
+
+  void SetChild(unsigned i, BTreeNode* child) {
+    Children()[i] = child;
+  }
+
+  struct SearchResult {
+    uint32_t index : 31;
+    uint32_t found : 1;
+  };
```

Comment:
Only only use it to return result from functions and never place it anywhere besides the stack, surprised you use bitfields here 😵

Reply:
you are right.